### PR TITLE
Add routing rules engine feature

### PR DESCRIPTION
**Pull request by hecris**
Currently, presto-gateway routes requests to backends by reading the `X-Trino-Routing-Group` header. 

This PR introduces a routing rules engine feature, allowing users to define custom logic to route queries. These rules are separated from application logic and kept in a config file, allowing for dynamic rule changes. These rules can use any method of a `HttpServletRequest` object, including get headers (https://trino.io/docs/current/develop/client-protocol.html)

### CODE REVIEW

**Comment by riteshvaryani** [comment]
> Currently, presto-gateway routes requests to backends by reading the `X-Trino-Routing-Group` header.
> 
> This PR introduces a routing rules engine feature, allowing users to define custom logic to route queries. These rules are separated from application logic and kept in a config file, allowing for dynamic rule changes. These rules can use any method of a `HttpServletRequest` object, including get headers (https://trino.io/docs/current/develop/client-protocol.html)

Change this to X-Presto-Routing-Group

File: README.md [hunk 1]
```diff
@@ -295,6 +295,201 @@ To graceful shutdown a Presto cluster without query losses, the steps are:
 
 To gracefully shutdown a single worker process, see [this](https://trino.io/docs/current/admin/graceful-shutdown.html) for the operations.
 
+## Routing Rules Engine
+By default, presto-gateway reads the `X-Trino-Routing-Group` request header to route requests.
+If this header is not specified, requests are sent to adhoc.
```
Comment (riteshvaryani) [inline README.md:300]: Change this "requests are sent to adhoc." to 
"requests are sent to default routing group(adhoc)"

File: README.md [hunk 2]
```diff
@@ -295,6 +295,201 @@ To graceful shutdown a Presto cluster without query losses, the steps are:
...
+In the condition, you can access the methods of a [HttpServletRequest](https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html) object called `request`.
+There should be at least one action of the form `result.put(\"routingGroup\", \"foo\")` which says that if a request satisfies the condition, it should be routed to `foo`.
+
+The condition and actions are written in [MVEL](http://mvel.documentnode.com/), an expression language with Java-like syntax.
+In most cases, users can write their conditions/actions in Java syntax and expect it to work. There are some MVEL-specific operators that could be useful though.
+For example, instead of doing a null-check before accessing the `String.contains` method like this:
+```yaml
+condition: "request.getHeader(\"X-Trino-Client-Tags\") != null && request.getHeader(\"X-Trino-Client-Tags\").contains(\"label=foo\")"
+```
+You can use the `contains` operator
```
Comment (riteshvaryani) [inline README.md:331]: This is a good callout as we discussed during the call!

**Comment by riteshvaryani** [comment]
LGTM! some minor comments and then push?!!

File: README.md (hunk 1 above)
Comment (hecris) [reply README.md]: done!

**Comment by hecris** [comment]
> LGTM! some minor comments and then push?!!

thanks! just addressed your comment

**Review by riteshvaryani — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -295,6 +295,201 @@ To graceful shutdown a Presto cluster without query losses, the steps are:
...
+```yaml
+condition: "request.getHeader(\"X-Trino-Client-Tags\") contains \"label=foo\""
+```
+If no rules match, then request is routed to adhoc.
+
+### Execution of Rules
+All rules whose conditions are satisfied will fire. For example, in the two example rules given above, a query with source `airflow` and label `special`
```
[the remark this answers is not in the document]
Comment (hecris) [reply README.md:338]: i was referring to the "airflow" and "airflow special" rules from above, maybe i should make it more clear

File: README.md [hunk 4]
```diff
@@ -295,6 +295,201 @@ To graceful shutdown a Presto cluster without query losses, the steps are:
...
+    else {
+      result.put(\"routingGroup\", \"etl\")
+    }"
+```
+
+### Enabling routing rules engine
+To enable routing rules engine, add the following lines to `gateway-ha-config.yml`.
+```
+routingRules:
+  rulesEngineEnabled: true
```
[the remark this answers is not in the document]
Comment (hecris) [reply README.md:490]: yup, changed instructions

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hecris) [reply README.md]: changed, let me know if it's still unclear

[the remark this answers is not in the document]
Comment (riteshvaryani) [reply README.md]: looks good

**Pull request merged**